### ISSUE TITLE
User-defined options on pull-down menu

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mintty/mintty#100]
```
It would be very useful to have some user-defined options on the
right-click pull-down menu.  For example, one option I would add is similar
to an option I use in an Ubuntu window, sending a highlighted strip as a
link to be opened in my browser.  I might add another option like sending a
highlighted strip to whois in my Cygwin window, etc.

Thanks.

Lester
```

Original issue reported on code.google.com by `LesterIn...@gmail.com` on 2 May 2009 at 2:47

**Comment by mintty** [comment]
The repository version now supports user-defined commands. See manual page for a description.
Please give it a try and provide feedback.

**Comment by mkq** [comment]
Thanks. There's still some work TBD before I switch back from urxvt to mintty :-)

1. I downloaded the master zip, ran make, viewed docs/mintty.1.html to find nothing about user commands. Took a while to find out that mintty.1.html contains less than mintty.1.
2. Config file formal: The first separator I tried was Ctrl+L which does not work. Switched to Ctrl+A, which works.
3. When using the full scrollback instead of a selection, we need a possibility to pass it to the command as stdin. I believe that it is impossible to properly quote %s in the configured command when used as an argument, because the scrollback will typically already contain both single and double quote characters and linefeeds.
4. As mentioned in #475, a page of scrollback would be nice.

**Comment by mintty** [comment]
Thanks for the feedback.
1. Currently, Makefile generates mintty.1.html only when building a release package; sorry. I'll look after that.
2. Ctrl+L: That's weird, I even had a crash. I'll check.
3. Yes, proper quoting is an issue here and I'm considering to put all information into environment variables. I think that would be the best solution, no escape problems. Stdin might be an option but only for one stream, thus inconsistent with further information provided (such as window title, name and working directory of foreground program, whatever...). Also, I now use popen() as an easy implementation mechanism to get the output back into mintty, but there is no bidirectional popen() available and it would need to be implemented the painful way, with lowest level fork and file descriptor tricks...
4. I don't see a use case for accessing precisely one "page". What would be the definition of "the page" and why would you want it?

**Comment by mintty** [comment]
2. solved: it's due to the `isspace` macro that skips white-space when the config file is parsed. Ctrl+L is a "Form Feed" and thus skipped. Since this is generic in options handling, I'd rather not take a risk in changing anything here but mention it in the manual.

**Comment by mkq** [comment]
3. You're right, environment variables (plus extra info) would be handy.
4. What I mean with "page" is the currently visible part (i.e. the number of lines would be the current window height if no line is wrapped; fewer lines otherwise). Not necessarily the bottommost content; rather scroll position should be taken into account.
The use case: After executing a command whose (e.g. tabular) output looks messed up due to line wrapping, I'd often only need a small part of the scrollback in order to view it in less. 
A page is not necessarily the best size, just the most straight-forward which came to mind. 
If mintty can tell the difference between prompts and the rest, maybe it could pass the last non-empty command output instead. (But then, if we have scrolled back so that multiple previous commands' output is shown, which should be used? The latest fully visible, maybe. E.g. seeing end of command A; all of B; all of C; beginning of D: => use C.)

**Comment by mintty** [comment]
3. Thanks for confirming, so be it. Which additional/extra info would be useful?
4. That sounds too complicated. The current screen is a good terminal-related concept, also used already in DEC terminal printing features (although screen printing is not currently implemented in mintty, just cloned output printing). One other option might be the last marked section, in a similar manner as discussed in #569. You could then include markers in your prompt and would get the most recent command output, more or less. However, I'm still lacking an implementation of finding such a mark.

**Comment by mkq** [comment]
> 2. solved: it's due to the isspace macro that skips white-space when the config file is parsed. Ctrl+L is a "Form Feed" and thus skipped. Since this is generic in options handling, I'd rather not take a risk in changing anything here but mention it in the manual.

I overlooked that post yesterday. Thanks, I agree documenting it is sufficient.

3. Additional info variables I can think of:
  - +1 for your window title suggestion (when I start my custom command in separate window, it's good give it a similar title)
  - The line number of the topmost visible line (e.g. for `less +${LINENO}g`)
  - (Just guessing:) Maybe which `/dev/pty*` this mintty is connected to? (Would that allow the user command to read interactive input typed into the mintty window or do any other useful things?)
  - Maybe (if not too slow) a variable with the same value as the main scrollback contents variable, but this time including escape sequences (colors etc.). Or maybe only one variable, but the config specifies on a per-command basis whether or not to include escapes.

4. Sounds good. (And similar to an idea I had in the meantime: No need to safely discriminate between what was printed as a prompt. If there was a regex option what should separate scrollback chunks, the user could set it so it more or less only matches his prompt). 
But IIUC when using marks instead of detecting prompts, the question remains what the right (considering the scroll position) chunk is when multiple are visible (some completely; some partially maybe). Was that the part you find too complicated?

**Comment by mintty** [comment]
* topmost line number with respect to beginning of buffer, I suppose? Don't know whether that helps in the presence of wrapped lines...
* /dev/pty*: yes and no. (I was thinking intially to even connect the command directly to the pty, as an option.); the problem is that you would have two programs competing for input at the same pty. Because the shell cannot know that there's a forked program also wanting to read from its pty, it would continue to at the same, most likely resulting in unpleasant and undesirable behaviour. Of course mintty input could in theory _temporarily_ be redirected to the application, but... (who should implement *that*?).
* ESC sequences would have to be regenerated from the internally stored text attributes. While something similar happens with the "Copy as rich text" option, it would require some unpleasant hacking to implement which I don't feel like doing right now...
4. What I found complicated is to find a convining definition of the "part of interest" following your ideas. "Visible screen" or "last command output" are two fairly simple concepts to that aim, although quite different. And for "last command output" there is still some implementation effort to expect, again complicated:( , otherwise I would have implemented #569 meanwhile.

**Comment by mkq** [comment]
> topmost line number with respect to beginning of buffer, I suppose? Don't know whether that helps in the presence of wrapped lines...

Yes. My assumption (and impression of how it currently works with %s) was that the user command gets logical lines as they were initially printed, not wrapped lines. Then the line number variable could be used to position less at the same line the currently scrolled-to page starts at. 
If it gets wrapped lines however, that would defeat my main use case (using less to make mintty a "temporarily horizontally scrolling" terminal).

OK, the pty thing was just an idea / brainstorming. I had no plans what to do with it anyhow.

**Comment by mintty** [comment]
I've uploaded the revised version. See manual, plus this:
> the command is invoked and its standard output is pasted into the terminal
>, applying bracketed-paste mode if enabled.

**Comment by mintty** [comment]
In combination with #569, mintty now also provides the last command output, in variable MINTTY_COMMAND.
Please check.

**Comment by mkq** [comment]
Great! Thank you very much. I notices some issues though (with version fffdf62):

$MINTTY_SCREEN starts at the correct scroll position, but ends at the overall last line.

$MINTTY_COMMAND used inside the prompt gives me the command I'm about to start or the currently executing if that is running long enough. So instead, I print ESC[?7711h in zsh's preexec function. That gives the last finished command's output as I wanted, except its first line is somehow always missing.

$MINTTY_SELECTION: With something selected, there's a strange side effect: Mintty's displayed selection changes to a region of full rows. I think it's from the last ESC[?7711h to the last line (regardless of whether the actual selection is in that range or somewhere nearer to the top).
What I mean by "displayed selection" is: It just looks as if I had selected those lines; what is copied to the clipboard is correct (what I actually selected).
BTW, the variable's value is correct in both situations (with or without selection).

The mintty / zsh config I have tested with:
```
UserCommands=^Atest:printf '===sel===\n%s\n===cmd===\n%s\n' "$MINTTY_SELECTION" "$MINTTY_COMMAND">/tmp/mintty.log
```
```
tputReset="%{$(tput sgr0)%}"
PS1=$(strjoin '' \
	$'\n' \
	'%{ESC[48;5;7m%}' \
	'%(0?..' '%{ESC[1m%}%{ESC[38;5;1m%}' '[status %?] ' "$tputReset" '%{ESC[48;5;9m%}' ')' \
	'$(histLockWarn)' "$tputReset" '%{ESC[48;5;7m%}' '%~' \
	'$(reset_color= git_super_status) ' \
	'%{ESC[38;5;12mESC[1m%}' '%{ESC[K%}' \
	"$tputReset" $'\n> '
)

preexec() {
	local cmd
	print -n 'ESC[?7711h'
	if [ $#1 -le 40 ]; then cmd=$1; else cmd="${1[1,37]}..."; fi
	cmd=$(perl -ple 's{([^[:graph:] ])}{sprintf("\\%03o", ord($1))}ge' <<< "$cmd")
	print -n '\e]0;'
	print -rn -- "$cmd (zsh "
	if [ ${+zshtitle} -ne 0 ]; then print -Pn "$zshtitle"; else print -Pn "$$"; fi
	print -n ")\\a"
	lastStartTime=$SECONDS
}
```
Escape characters replaced with "ESC".
`strjoin ''` just joins the args into a single line of output.
histLockWarn prints nothing if all is well.
git_super_status printed nothing in my tests, because I was not in a git directory.

**Comment by mintty** [comment]
Thanks for the feedback.

> $MINTTY_SCREEN starts at the correct scroll position, but ends at the overall last line.

That's its purpose, basically to provide the whole screen contents, just extend it backwards when scrolled back.

> $MINTTY_COMMAND used inside the prompt gives me the command I'm about to start or the currently executing if that is running long enough. So instead, I print ESC[?7711h in zsh's preexec function. That gives the last finished command's output as I wanted, except its first line is somehow always missing.

What do you mean by MINTTY_COMMAND inside the prompt? It's only set for user-defined commands, not for the shell.
I'm not familiar with preexec but I guess it's output appears after the command line already, so on the first command output line, which is thereby marked as "prompt line" (that's how the feature is designed, maybe I can describe that clearer). That would explain the effect.

> $MINTTY_SELECTION: With something selected, there's a strange side effect: Mintty's displayed selection changes to a region of full rows. I think it's from the last ESC[?7711h to the last line (regardless of whether the actual selection is in that range or somewhere nearer to the top).
What I mean by "displayed selection" is: It just looks as if I had selected those lines; what is copied to the clipboard is correct (what I actually selected).
BTW, the variable's value is correct in both situations (with or without selection).

The contents of MINTTY_SELECTION isn't related to the scroll markers. But it obviously affects visual highlighting (and maybe related there, I'll check) which is not intended.

> The mintty / zsh config I have tested with: ...
I hope you don't suppose me to analyse your complex prompt setup...

**Comment by mkq** [comment]
$MINTTY_SCREEN: OK, thanks for clarifying.

> What do you mean by MINTTY_COMMAND inside the prompt? It's only set for user-defined commands, not for the shell.

Sorry, I didn't make sense there. I meant: If ESC[?7711h is inside the prompt, $MINTTY_COMMAND gives me ...
You are right about preexec, as demonstrated here:
```
> preexec() { print -n '[pre]'; }
> date; date
[pre]So, 26. Feb 2017 16:11:24
So, 26. Feb 2017 16:11:24
>
```
So now I understand the feature works as designed, and it is really useful for scrolling. But with the marker inside the prompt, the variable will be empty often (most of the time?). I.e. every time the shell is prompting. Could the variable be changed to contain the last non-empty marker-delimited part? So that would still be the current output if a command is running, and the last command's output otherwise.

On second thought, I think I would rather use separate variables for selection and full contents (say MINTTY_SELECTION and MINTTY_TEXT). I myself will probably only use the non-selection case, so the current behavior is sufficient for me. But I guess others might want commands to work on the selection only (and do nothing if nothing is selected). Currently, they would need to implement some guess to accomplish that. Like if $MINTTY_SELECTION contains \<something which looks like their prompt\> or is more than \<n\> characters or lines long, it's probably not a selection. If these features were split into two variables, the current behavior would still be easy to get with ${MINTTY_SELECTION:-$MINTTY_TEXT}.

I was about to repeat my wish from #475 to mark the keyboard shortcuts in the menu items with "&", but that already works. Nice! This seems to be undocumented.

> \> The mintty / zsh config I have tested with: ...
> I hope you don't suppose me to analyse your complex prompt setup...

No, I just included that in case you wanted to use parts of it to reproduce. And I was too lazy to work out a minimal config. Of course, I will if it helps.

**Comment by mintty** [comment]
A command can well have empty output, and it that case I think it is proper to provide empty MINTTY_COMMAND. So to be consistent, if you nervously hit Enter after your previous command :) , you'll have spoiled the feature and be out of luck. Sorry, but that's consistent with the definition of "last command output".
Or do you have a multi-line prompt? That's currently not supported, and the mark should be set in only 1 line in that case.
Handling selection and full scrollback buffer separately may be a good idea, I just wanted to reduce the amount of text to be handled...
Looking for better names for MINTTY_COMMAND and (then new) MINTTY_TEXT.

**Comment by mkq** [comment]
Yes, my prompt is three lines long. I moved the ESC[?7711h to the last one (and removed it from preexec), and all seems to work. Cool! Thanks again for your support.

**Comment by mintty** [comment]
Yes, last line of prompt I should have said, or the subsequent prompt lines will be taken for command output.
I could add support for multi-line prompt. You would have to set secondary marks in the upper lines to exclude them from the notion of last command output.

**Comment by mkq** [comment]
That would be nice to have.

But currently, I plan to use ..._SCREEN and ..._SELECTION (or its successor for non-selection), but probably not MINTTY_COMMAND. I was trying to do QA; turned out to be rather wasting your time - sorry. At least it showed where the documentation could be more fool-proof. :-)

**Comment by mintty** [comment]
Now supporting multi-line prompts (you can set a secondary marker to exclude the upper prompt lines from the previous command output).
Changed variables:
MINTTY_SELECT (shortened) selection
MINTTY_BUFFER full scrollback buffer
MINTTY_OUTPUT previous command output

**Comment by mintty** [comment]
Released with 2.7.5.

**Comment by mkq** [comment]
Unfortunately, too late for the release, but: The documentation still says MINTTY_SELECTION in a note and the examples.

And off topic: "the Windows console window that Cygwin uses by default" in the doc is not up to date.
